library/core/class.model.php

```diff
@@ -961,4 +961,42 @@ public static function editContentTimeout($data, &$timeLeft = 0) {
 
         return $canEdit;
     }
+
+    /**
+     * Locks a resource so that it can only be accessed one at a time.
+     *
+     * @param string $lockKey Cache key to be assigned.
+     * @param int $gracePeriod Period of time the key will stay valid.
+     * @return bool $hasMasterKey Whether a master key has been assigned.
+     */
+    protected static function buildCacheLock(string $lockKey, int $gracePeriod = 60):bool {
+        $hasMasterKey = null;
+
+        if (is_null($hasMasterKey)) {
+            // First instance will receive the $master key and lock the record.
+            $instanceKey = getmypid();
+            $masterKey = Gdn::cache()->add($lockKey, $instanceKey, [
```

Comment:
Are you sure it'll work like this? It looks like you're expecting it to return the cached value, but I don't think that'll be the case. For example, the `Gdn_Memcached` caching adapter's `add` method returns boolean values. You're comparing that result, stored as `$masterKey`, with the result of `getmypid`, which would be an integer. In PHP, this might coincidentally work on the first attempt, because `1234 == true` evaluates to `true` in PHP. Subsequent calls, even in the same process, would see the value of `$masterKey` being `false`, which would then lead to `$hasMasterKey` also being `false`. This is because the caching adapter's `add` method adds new cached values under a key __if it doesn't already exist__ and returns a success state *or does nothing and responds with a failure state if the cache key already exists.*

Reply:
This is an error in our original example code. Thus proving you need to double-check assumptions when copying code.